examples/auction/auction_cc.cpp

```diff
@@ -21,25 +22,96 @@
 
-static std::string SEP = ".";
-static std::string PREFIX = SEP + "somePrefix" + SEP;
+#define INITALIZED_KEY "__initalized"
+#define AUCTION_HOUSE_NAME_KEY "__auction_house_name"
+
+const std::string SEP = ".";
+const std::string PREFIX = SEP + "somePrefix" + SEP;
 
 // implements chaincode logic for invoke
-int invoke(const char* args,
-    uint8_t* response,
-    uint32_t max_response_len,
-    uint32_t* actual_response_len,
-    void* ctx)
+int init(
+    uint8_t* response, uint32_t max_response_len, uint32_t* actual_response_len, shim_ctx_ptr_t ctx)
 {
-    LOG_DEBUG("AuctionCC: +++ Executing auction chaincode invocation +++");
-    LOG_DEBUG("AuctionCC: \tArgs: %s", args);
+    // Note: we could have got here not only during instatiation but also due to an upgrade.
+    // we do allow this, so don't check status of _initialized here
 
+    LOG_DEBUG("AuctionCC: +++ Executing auction chaincode init+++");
     std::vector<std::string> argss;
     // parse json args
-    unmarshal_args(argss, args);
+    get_string_args(argss, ctx);
+
+    LOG_DEBUG("AuctionCC: Args: %s",
+        (argss.size() < 1
+                ? "(none)"
+                : std::accumulate(std::next(argss.begin()), argss.end(), argss[0],
+                      [](std::string a, std::string b) { return (a + std::string(", ") + b); })
+                      .c_str()));
+
+    const char* _auction_house_name = argss[0].c_str();
+    put_state(
+        AUCTION_HOUSE_NAME_KEY, (uint8_t*)_auction_house_name, strlen(_auction_house_name), ctx);
+
+    bool _initialized = true;
+    put_state(INITALIZED_KEY, (uint8_t*)&_initialized, sizeof(_initialized), ctx);
+
+    *actual_response_len = 0;
+    LOG_DEBUG("AuctionCC: +++ Initialization done +++");
+    return 0;
+}
+
+// implements chaincode logic for invoke
+int invoke(
+    uint8_t* response, uint32_t max_response_len, uint32_t* actual_response_len, shim_ctx_ptr_t ctx)
+{
+    bool _initialized;
+    const char* _auction_house_name;
+    char _auction_house_name_buf[128];
 
-    std::string function_name = argss[0];
-    std::string auction_name = argss[1];
+    uint32_t init_len = -1;
+    get_state(INITALIZED_KEY, (uint8_t*)&_initialized, sizeof(_initialized), &init_len, ctx);
+    if ((init_len == 0) || !_initialized)
+    {
+        _initialized = false;
+        _auction_house_name = "(uninitialized)";
+    }
+    else
+    {
+        uint32_t ahn_len = -1;
+        get_state(AUCTION_HOUSE_NAME_KEY, (uint8_t*)_auction_house_name_buf,
+            sizeof(_auction_house_name_buf) - 1, &ahn_len, ctx);
+        if (ahn_len == 0)
+        {
+            _auction_house_name = "(uninitialized)";
+        }
+        else
+        {
+            _auction_house_name_buf[ahn_len + 1] = '\0';
+            _auction_house_name = _auction_house_name_buf;
+        }
+    }
+
+    LOG_DEBUG(
+        "AuctionCC: +++ Executing '%s' auction chaincode invocation +++", _auction_house_name);
+
+    if (!_initialized)
```

Comment:
This is tricky! So this code only works if init is called. This means that this chaincode only works with lifecycle v2 with init enabled. 

Also, ... invocation of init should be enforced by the peer. If the peer is correct, this code will never be reached?!

Reply:
not necessarily, init could also have failed?  In any case, if with lifecycle v2 then you definitely must enforce such initialized aspects yourself in the code if your state really depends on it ?